packages/crypto/src/crypto.ts

```diff
@@ -0,0 +1,498 @@
+  const hexString = num.toString(16);
+  if (hexString.length > length) {
+    throw new Error(
+      `number cannot fit in a hex string of ${length} characters`
+    );
+  }
+  return hexString.padStart(length, "0");
+};
+
+/**
+ * Uncompress a raw public key.
+ */
+
+const uncompressRawPublicKey = (rawPublicKey: Uint8Array): Uint8Array => {
+  // point[0] must be 2 (false) or 3 (true).
+  // this maps to the initial "02" or "03" prefix
+  const lsb = rawPublicKey[0] === 3;
+  const x = BigInt("0x" + uint8arrayToHexString(rawPublicKey.subarray(1)));
+
+  // https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf (Appendix D).
+  const p = BigInt(
+    "115792089210356248762697446949407573530086143415290314195533631308867097853951"
+  );
+  const b = BigInt(
+    "0x5ac635d8aa3a93e7b3ebbd55769886bc651d06b0cc53b0f63bce3c3e27d2604b "
+  );
+  const a = p - BigInt(3);
+
+  // Now compute y based on x
+  const rhs = ((x * x + a) * x + b) % p;
+  let y = modSqrt(rhs, p);
+  if (lsb !== testBit(y, 0)) {
+    y = (p - y) % p;
+  }
+
+  if (x < BigInt(0) || x >= p) {
+    throw new Error("x is out of range");
+  }
+
+  if (y < BigInt(0) || y >= p) {
+    throw new Error("y is out of range");
+  }
+
+  var uncompressedHexString = "04" + bigIntToHex(x, 64) + bigIntToHex(y, 64);
+  return uint8arrayFromHexString(uncompressedHexString);
+};
+/**
+ * Compute the modular square root using the Tonelli-Shanks algorithm.
+ */
+
+const modSqrt = (x: bigint, p: bigint): bigint => {
+  if (p <= BigInt(0)) {
+    throw new Error("p must be positive");
+  }
+  const base = x % p;
+
+  // Check if p % 4 == 3 (applies to NIST curves P-256, P-384, and P-521)
+  if (testBit(p, 0) && testBit(p, 1)) {
+    const q = (p + BigInt(1)) >> BigInt(2);
+    const squareRoot = modPow(base, q, p);
+    if ((squareRoot * squareRoot) % p !== base) {
+      throw new Error("could not find a modular square root");
+    }
+    return squareRoot;
+  }
+
+  // Other elliptic curve types not supported
+  throw new Error("unsupported modulus value");
+};
+
+/**
+ * Compute the modular exponentiation.
+ */
+const modPow = (b: bigint, exp: bigint, p: bigint): bigint => {
+  if (exp === BigInt(0)) {
+    return BigInt(1);
+  }
+  let result = b;
+  const exponentBitString = exp.toString(2);
+  for (let i = 1; i < exponentBitString.length; ++i) {
+    result = (result * result) % p;
+    if (exponentBitString[i] === "1") {
+      result = (result * b) % p;
+    }
+  }
+  return result;
+};
+
+/**
+ * Test if a specific bit is set.
+ */
+const testBit = (n: bigint, i: number): boolean => {
+  const m = BigInt(1) << BigInt(i);
+  return (n & m) !== BigInt(0);
+};
+
+/**
+ * Generate a random Uint8Array of a specific length.
+ */
+const randomBytes = (length: number): Uint8Array => {
+  const array = new Uint8Array(length);
+  for (let i = 0; i < length; i++) {
+    array[i] = Math.floor(Math.random() * 256);
+  }
+  return array;
+};
```

Comment:
This isn't secure because these random bytes are from a PRNG, not a CSPRNG. I think we need to find a way to use  the system's CSPRNG using e.g. https://github.com/LinusU/react-native-get-random-values

Because we're using `randomBytes` to create a private key it's not okay to rely on a PRNG unfortunately :(

Reply:
Good call!